Sources/Account/Balance.swift

```diff
@@ -5,95 +5,146 @@
+        //
         // amountLowMobDec = amountLow / 10^12
         // amountHighMobDec = amountHigh * 2^64 / 10^12
         //
+        // >> where 10^12 is 10^(significantDigits)
+        //
         // amountMobDec = amountLowMobDec + amountHighMobDec
         //
+        // >> Now expand amountLowDec & amountHighDec to their Integer & Fractional parts
+        //
+        // >> amountLowDec = amountLowMobInt.amountLowPicoFrac
+        //
         // amountLowMobInt = floor(amountLow / 10^12)
         // amountLowPicoFrac = amountLow % 10^12
-
+        //
+        // >> amountHighDec = amountHighMobInt.amountHighPicoFrac
+        //
         // amountHighMobInt = floor((amountHigh * 2^64) / 10^12)
+        //
+        //                  >> factor out common 2^12 for now
+        //                  = floor((amountHigh * (2^52 * 2^12)) / (5^12 * 2^12)
+        //
+        //                  >> bitshift by 52 (same as multiply by 2^52)
+        //                  >> ... and bitshift number == (64 - significantDigits)
         //                  = floor((amountHigh << 52) / 5^12)
+        //
         // amountHighPicoFrac = (amountHigh * 2^64) % 10^12
-        //                   = ((amountHigh << 52) % 5^12) << 12
+        //
+        //                  >> re-apply the 2^12
+        //                  = ((amountHigh << 52) % 5^12) << 12
         //
         // amountPicoFracCarry = floor((amountLowPicoFrac + amountHighPicoFrac) / 10^12)
         //
         // amountMobInt = amountLowMobInt + amountHighMobInt + amountPicoFracCarry
         // amountPicoFrac = (amountLowPicoFrac + amountHighPicoFrac) % 10^12
 
-        let (amountLowMobInt, amountLowPicoFrac) = { () -> (UInt32, UInt64) in
-            // 10^12 = 1_000_000_000_000
-            let mobParts = amountPicoMobLow.quotientAndRemainder(dividingBy: 1_000_000_000_000)
-            return (UInt32(mobParts.quotient), mobParts.remainder)
+        
+        let significantDigits = tokenId.significantDigits
+        
+        let divideBy = UInt64(pow(Double(10), Double(significantDigits)))
+        let (amountLowMobInt, amountLowFrac) = { () -> (UInt64, UInt64) in
+            let mobParts = amountLow.quotientAndRemainder(dividingBy: divideBy)
+            return (UInt64(mobParts.quotient), mobParts.remainder)
         }()
 
-        let (amountHighMobInt, amountHighPicoFrac) = { () -> (UInt32, UInt64) in
-            // Intermediary = base of 5^-12 MOB
-            let amountHighIntermediary = UInt64(amountPicoMobHigh) << 52
-            // 5^12 = 244_140_625
-            let mobParts = amountHighIntermediary.quotientAndRemainder(dividingBy: 244_140_625)
-            return (UInt32(mobParts.quotient), mobParts.remainder << 12)
+        let (amountHighMobInt, amountHighFrac) = { () -> (UInt64, UInt64) in
+            let amountHighIntermediary = UInt64(amountHigh) << (64 - significantDigits)
+            let factored = UInt64(pow(Double(5), Double(significantDigits)))
```

Comment:
thoughts on how we could compute factored exactly once per tokenId (same for divideBy)?

Reply:
I think this is a premature optimization, if there is evidence that these calculation are causing noticeable slowdowns, id be willing to change my mind.

Reply:
creating `Balance` structs gets slow at > 1000, I have performance tests now, we should re-visit this when the number of available tokens exceeds 100, as that may be a point at which it affects user's perceived performance.